### ISSUE TITLE
Webpack client does not throw error code when build failed

### ISSUE THREAD

**Issue by happypoulp** [webpack/webpack#708]
Hi,

I'd like to know when a build failed using the webpack client (because of a module not found for example). Right now, it seems that the client is outputting the error to the console but the status code of the webpack client process is 0 (success).

Is it the normal behavior? I can imagine that when using the watch option we don't want the process to end on a build failure but without the option, could the client terminate with a status != 0?

I'd like to avoid to search for "ERROR" inside the client output to determine that the build failed.

Thanks.

**Comment by happypoulp** [comment]
I found a way to get the desired result using the 'done' plugin:

```javascript
// ...
plugins: [
    // ...
    function()
    {
        this.plugin("done", function(stats)
        {
            if (stats.compilation.errors && stats.compilation.errors.length)
            {
                console.log(stats.compilation.errors);
                process.exit(1);
            }
            // ...
        });
    }
    // ...
],
// ...
```

Is this the correct wat to go?

Thanks.

**Comment by jhnns** [comment]
That is valid approach (you could also just throw an error :)).

And yes, I agree with you: webpack *should* exit with error code 1 when a build error occurred and watch is not active. I remember some discussion with @sokra about this topic ^^

**Comment by happypoulp** [comment]
Ok, thanks for the quick reply!

I'll go with this solution. However, going through a throw new Error(...) produce a little bit more output (with the stack trace of the error) that I do not need.

For the records, here is the version that avoids exiting the client process when the --watch option is present:

```javascript
// ...
plugins: [
    // ...
    function()
    {
        this.plugin("done", function(stats)
        {
            if (stats.compilation.errors && stats.compilation.errors.length && process.argv.indexOf('--watch') == -1)
            {
                console.log(stats.compilation.errors);
                process.exit(1); // or throw new Error('webpack build failed.');
            }
            // ...
        });
    }
    // ...
],
// ...
```

I'll leave the issue open for now since it would be nice to have webpack client exit with error code 1 on build failure. Feel free to cancel it if you think the present solution is enough.

**Comment by sokra** [comment]
`--bail`

**Comment by happypoulp** [comment]
@sokra Please could you detail the use of the --bail option?

I tried: ./node_modules/.bin/webpack --progress --colors --config --bail conf/webpack.config.js

the output seems to be truncated but the client does not exit with an error code.

**Comment by sokra** [comment]
Try `./node_modules/.bin/webpack --progress --colors --config conf/webpack.config.js --bail`

**Comment by happypoulp** [comment]
Works like charm, thanks!

**Comment by nelix** [comment]
I would not exactly say this is well documented, but vaguely referenced.

**Comment by Turbo87** [comment]
@sokra could you point out why `--bail` is not the default behavior in non-watch mode?

**Comment by sokra** [comment]
I can change it with the next major version...

**Comment by Turbo87** [comment]
@sokra  :+1: for that

**Comment by marr** [comment]
Closed but this is still not default behavior..

**Comment by sokra** [comment]
> Closed but this is still not default behavior..

It's default in the `webpack-2` branch. That's the next major version. It's a breaking change...

**Comment by aindlq** [comment]
@sokra I've noticed that --bail flag doesn't work in 1.9.*
And it seems some other users had similar experience -http://stackoverflow.com/a/29510310/124722

**Comment by TiddoLangerak** [comment]
As a temporary workaround I've created a small plugin based on the code from @happypoulp above that will exit with error code 1 when the build fails in single-run mode. You can find it [here](https://github.com/TiddoLangerak/webpack-fail-plugin).

**Comment by igl** [comment]
Can errors always be written to stderr?
I was just trying to send webpacks stdout to /dev/null because i do not care about the build report.
However without --bail stderr seems to remain empty.

**Comment by kilianc** [comment]
I don't understand why --bail is related to proper exit code. If there is an error exit code needs to be > 0, bail or not, and bail should not be default behavior.

**Comment by alexch** [comment]
@kilianc :+1: 

This discussion (and code) is confounding at least 3 different cases:

1. return a non-zero error code on failure
2. exit immediately on failure (is this what `--bail` is supposed to mean?)
3. how verbose should error output be

Regardless of 2 and 3, if an error occurred during the build, then when webpack exits, it **must** exit with a non-zero error code. Otherwise this is unusable in a continuous build script. 

That's not a breaking change. It's fixing a major bug.

Whether `--bail` should be default behavior is also interesting but a completely different question.

**Comment by alexch** [comment]
related: 
https://github.com/andreypopp/typescript-loader/issues/21 
https://github.com/webpack/webpack/issues/708
https://github.com/webpack/webpack/issues/711

**Comment by cerisier** [comment]
100% agree with @alexch 
Just had major headache finding out why deployed build was broken although CI showed all build passed...
What is the state on this ?
Can we help in any way ?

**Comment by asagage** [comment]
:+1: @alexch

**Comment by jhnns** [comment]
> I can change it with the next major version...

@sokra what's the state for webpack 2 regarding this issue? Wouldn't a non-zero exit code be good?

**Comment by giorgiosironi** [comment]
This is a serious bug, :+1:  to @cerisier and @alexch. You're going to deploy a bundle containing
```
   (function webpackMissingModule() { throw new Error("Cannot find module \"...\""); }());                                                             
```
without finding out until tests are executed against it.

**Comment by cjb** [comment]
@giorgiosironi Yep, just did exactly this.  :(

**Comment by nathancahill** [comment]
Just had this happen to me. +1 for non-zero exit code.

**Comment by StephanBijzitter** [comment]
Same happened here... why is this not fixed @sokra ?

**Comment by JustFly1984** [comment]
had the same issue with webpack exiting without proper exit code

fixed this by adding to webpack config:

```
plugins: [
    function() {
      this.plugin("done", function(stats) {
        if (stats.compilation.errors && stats.compilation.errors.length &&
process.argv.indexOf("--watch") == -1) {
          console.log(stats.compilation.errors);
          process.exit(1);
        }
      });
    }]

2016-03-29 21:55 GMT+08:00 Stephan Bijzitter <notifications@github.com>:

> Same happened here... why is this not fixed @sokra
> <https://github.com/sokra> ?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/webpack/webpack/issues/708#issuecomment-202904903>
>

**Comment by kemsky** [comment]
This is partial fix, logging stats.compilation.errors is ugly and has various issues when used with karma or gulp. I wish there was an appropriate option for this.

**Comment by igl** [comment]
@kemsky Printing the errors into stderr using `console.error()` does not help it?

**Comment by kemsky** [comment]
@igl, error objects contain different sets of properties depending on context (karma, gulp, standalone), error messages are not properly formatted in this case, you loose original webpack messages, because process is interrupted before webpack writes to stdout, configuration become bloated, gulp task can not exit correctly (forget to signal async completion?) and etc.

**Comment by PhiLhoSoft** [comment]
@aindlq It is also my experience with Webpack 1.12 on Windows 10. `echo %errorlevel%` shows 0 if I use the command line `webpack --bail --progress --profile` and have an error (eslint-loader reporting a missing semi-colon, for example).

@TiddoLangerak Your plugin works fine, thank you. In the above case, the reported errorlevel is 1.
I was surprised I only had to put the name of the plugin in the list of plugins. All other plugins I used in my configuration use a `new PluginName()` instead.

About the above remarks on `console.log(errors)`, _webpack-fail-plugin_ just exits and the errors are correctly reported on the console.

**Comment by razorcd** [comment]
If you start the `webpack` build from a `gulp` task and `webpack` fails then `gulp` will still exit with `0`. Even if using the `--bail` flag.
```
gulp.task('build-js', function () {
    webpack(webpackConfig).run(...);
});
```

Solution: use [webpack-fail-plugin](https://www.npmjs.com/package/webpack-fail-plugin) and the `gulp` task will exit with `1`.

**Comment by svarnoi420** [comment]
This "my" solution from here [http://webpack.github.io/docs/node.js-api.html#error-handling](http://webpack.github.io/docs/node.js-api.html#error-handling)

And I don't used **webpack-fail-plugin**.

I used **gulp-util** for format webpack log to gulp format.

```
const webpack = require('webpack');
const gutil   = require('gulp-util');

let webpackConfig = require('../webpack.config.js');
let statsLog      = {
	colors: true,
	reasons: true
};

gulp.task('scripts', (done) => {
	// run webpack
	webpack(webpackConfig, onComplete);

	function onComplete(error, stats) {
		if (error) {
			let fatalError = new gutil.PluginError('webpack', error);

			done(fatalError);
		} else if ( stats.hasErrors() ) {
			let softError = new gutil.PluginError('webpack', stats.toString(statsLog));

			done(softError);
		} else {
			gutil.log('[webpack]', stats.toString(statsLog));

			done();
		}
	}
});
```

**Comment by cjwilliams** [comment]
Are there plans/desires to make non-zero exit codes standard? If so, should we not re-open this issue?

**Comment by wonderbeyond** [comment]
> I can change it with the next major version...

It's something must! error without status=0 is evil!

**Comment by jhnns** [comment]
I agree with all of you. I don't understand what the rationale behind the current behavior was.

However, since we're about to finish webpack 2 (which will exit with a non-zero exit code by default), we won't add any non-critical things to webpack 1 anymore.

**Comment by StephanBijzitter** [comment]
That's only two years too late...

**Comment by jhnns** [comment]
If you want to speed up development, [funding](https://opencollective.com/webpack) helps more than complaining.

**Comment by StephanBijzitter** [comment]
That would not fix the priorities.

Somehow, webpack 2 is getting more priority than this bug. This is webpack, a command-line build tool that always exits with status 0, which makes it largely useless unless you fix it manually after googling. If you guys had more money, perhaps webpack 2 would arrive a month sooner, but this bug would still be in the webpack 1 releases.

I'd even just fix it myself and submit a pr, but you wouldn't even want to merge a bugfix as you believe it's a breaking change. No, it's not: it's fixing a broken tool.

**Comment by StevenIseki** [comment]
Just use the `--bail` option until webpack 2 comes out as sokra mentioned, it works well for me, and allows me to see errors in CI builds.

**Comment by PhiLhoSoft** [comment]
@StevenIseki Looks like you haven't read the whole thread...

**Comment by urbanhusky** [comment]
I actually think that this isn't a bug. During development, with hot module replacement, you wouldn't want the webpack dev server to crash whenever you've created invalid TypeScript etc.

Certainly, you would want a non-zero error code for production builds (which do not utilise the webpack-dev server). Personally, I'm fine with adding the workaround/hack to my production configuration.

**Comment by tao12345666333** [comment]
Thanks all.

**Comment by sepo-one** [comment]
`--bail` does not help when throwing an error from loader

**Comment by blordpluto** [comment]
It looks like the last comment in this thread is still an issue. I have `--bail` set to true. A CSS loader is erroring. And a CI build is failing without aborting. Thus a false positive for success of the build. This is very confusing for CI. Has this really not been addressed for three years? 🤔 

I would like to be able to set an initial loader for single-run compile that catches any errors from the subsequent loaders, and then terminates the entire process. Something similar to the code at the top of this thread. But I don't see any way to hook into other loader's errors in the loader interface. Thanks in advance for any help.

**Comment by Yankovsky** [comment]
Throwing error from loader results in successful build with warning.
![https://ibb.co/8xrvBfx](https://ibb.co/8xrvBfx)

**Comment by sokra** [comment]
It's a warning so it doesn't result in an exit code.

But it probably shouldn't be a warning.

**Comment by Yankovsky** [comment]
Seems like if schema validation errors is wrapped in the same warning.
![http://s.csssr.ru/U02D248T6/2020-06-02-15-47-37-m4mqe_.jpg](http://s.csssr.ru/U02D248T6/2020-06-02-15-47-37-m4mqe_.jpg)

**Comment by Yankovsky** [comment]
@sokra Do you need help with that issue? Example repo?

**Comment by sokra** [comment]
If you want to you can send a PR. The problem is that elements in a "ContextModule" are considered as optional. That should be changed for webpack 5.

Look for ContextElementDependency and ContextHelper

**Comment by sokra** [comment]
fixed by #11327
